### ISSUE TITLE
File Upload to Storage working

### ISSUE THREAD

**Issue by sglavoie** [datopian/gift-publisher#21]
* Background: https://github.com/datopian/gift-portal/issues/23

## Acceptance

* [x] The publisher app has working file upload **DONE**- https://github.com/datopian/gift-publisher/blob/0a4599ea11ac416146b3a516e4e72fa72399f30e/src/components/Upload/index.jsx#L134
* [x] Files are stored in cloud storage at gcs://{gift-data-bucket}/{prefix}/{dataset-id}/sha256 **DONE** - 
![UPLOAD](https://user-images.githubusercontent.com/20904032/104589571-5eaf4900-566a-11eb-83a6-9744142336b1.png)

  * [ ] (BONUS?) Files have header set so that on download the file is correctly named ...
  * [x] (BONUS?) Mimetype correctly set

TODO: check github has unique ids separate from name for a repo

```
{dataset-id} = github id (not repo name so that we can handle renames etc)
{prefix} = blobs
{gift-data-bucket} = this will get configured on giftless server
```

## Tasks

TODO: Estimate how long this issue would take.

* [x] Implement upload method in giftless-client-js **DONE** - https://github.com/datopian/giftless-client-js/blob/main/src/index.js
* [x] Add method in lib in publisher code for uploading based on https://github.com/datopian/ckan-client-js/blob/master/lib/index.js#L158-L166 **DONE**- https://github.com/datopian/giftless-client-js/blob/main/src/index.js
  * [x] Create function `pushBlob` **Note** we might not need this currently, since it is a wrapper over `lfsCLient.upload`
  * [ ] `const token = await this.getUploadAuthToken()` - this i going to call the authorize API thadeu implemented **NOTE** yet integrated into `gift-portal` to have access to this function.
* [ ] Tests ... (may be painful or need to be in cypress) **NOTE** - Test is written but not yet integrated into `github workflow` due to uploading of files to cloud storage per push while uploading (need to find a way of mocking the `lfsSerevr` in cypress). 
      Also, the integration test stoped at **DescribedYour Dataset** due to the selection of `Rich Types`; The HTML generated by the `Select` component is not that clear

## Analysis

To do direct to cloud upload from our client side app we need:

* Client (Browser / Code) = Gift Publisher (DataPub)
* Authz Server = Need to build ourselves (in CKAN ckanext-authz-service)
* Storage Access Token Service = giftless
* Cloud Storage = Google / Amazon

Algorithm

* [x] Computes sha256 and gets other info (e.g. dataset id and size) that is used for composing upload storage request/location (where does this file live in storage?)
* [ ] Uses /api/authz endpoint to get token for giftless
* [x] Then goes to giftless server with token and gets upload url + headers
* [x] Uses url + headers to upload file to storage (using uppy? or datapub component?)
* [x] Inserts details into Resource on the Data package that is stored in the react app

## Existing ckanClient.js 

https://github.com/datopian/ckan-client-js
Example usage in datapub-nhs: https://github.com/datopian/datapub-nhs/blob/81b20860cf5f690bfa36893db9248a041ceefc04/src/components/Upload/index.jsx#L105

---

/cc @risenW (@steveoni may want to get going with @risenW's help - pair programming being an option)

**Comment by steveoni** [comment]
@sglavoie From the `Datapub-nhs` I can see that to upload is just a call to `pushblob` which is a wrapper over the `LfsClient.upload`, so my question is, is it `lfs-server` that handles cloud storage to store the resources?

**Comment by steveoni** [comment]
## Blocker
Experiencing `cross site origin` issue
```
Access to XMLHttpRequest at 'https://giftless-gift.herokuapp.com/gift-data/f.csv/objects/batch' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Comment by risenW** [comment]
@steveoni Have you tried using the cors library as a middleware in the routes?  See https://nextjs.org/docs/api-routes/api-middlewares

**Comment by steveoni** [comment]
this from `giftless-client-js` inside `gift-publisher`. I've try setting the header to contain `Allow cross site origin`

**Comment by sglavoie** [comment]
@steveoni We've discussed this issue earlier today and so we are aware of what's the next step for this issue, here's the solution: we need to support CORS on the Giftless server according to the instructions found here https://github.com/datopian/giftless#adding-cors-support

I will update you once the server is redeployed in the correct location (will have to move the private GitHub repo too) with the expected configuration. :+1:

**Comment by steveoni** [comment]
## NOTE

The file upload is now working, but the `cypress` is yet to be fully completed. **Reason**:
1. A need to mock the request to the `lfsServer` in cypress to prevent pushing to google cloud storage per push
2. A need to fully test the workflow; currently having issue testing second to the last workflow (Page)

There might be a need to create another issue to focus on this

**Comment by sglavoie** [comment]
WONTFIX: No immediate further action required. May consider adding Cypress tests at some point. Setting the Giftless token is covered in https://github.com/datopian/gift-portal/issues/138.

* File upload to storage is working.
* Files are properly named when downloaded.
